chapters/chap03-sec02.tex

```diff
@@ -124,18 +124,11 @@ \subsection{An efficient instantiation of \mimc{} primitives}\label{instantiatio
 \end{equation}
 
-For specific values of $\rCURVE$ (such as $\rBN$ for $\BNCurve$ or $\rBLS$ for $\BLSCurve$), it remains to choose concrete values of $\exponent{}$ and $\rounds{}$.
-
 \newcommand{\constraints}{\varstyle{constraints}}
 
-Note that small exponents $\exponent{}$ result in fewer constraints in the arithmetic circuit (see~\cref{zeth-protocol:statement}), while larger exponents can reduce the cost of Merkle tree operations on the contract (see~\cref{zeth-protocol:process-tx}). This is due to two factors, namely that exponentiation is cheaper to execute on a contract than in an arithmetic circuit, and that the number of rounds decreases with higher $\exponent{}$. For instance, choosing $\exponent{} = 7$ results in $365$ constraints and $\approx 20k$ gas while $\exponent{} = 31$ corresponds to $417$ constraints ($+15\%$) and $\approx 17k$ ($-10\%$) in gas consumption. Repeating the same process for different exponents, we observe roughly the same order of magnitude gain on the gas consumption and loss on the number of constraints.
+For specific values of $\rCURVE$ (such as $\rBN$ for $\BNCurve$ or $\rBLS$ for $\BLSCurve$), it remains to select concrete values of $\exponent{}$ and rounds, where $\rounds = \lceil \frac{\log_2 \rCURVE}{\log_2 \exponent{}} \rceil$).  These values influence the number of constraints in the arithmetic circuit (see~\cref{zeth-protocol:statement} for details of the statement) and the gas cost of Merkle tree operations on the contract (see~\cref{zeth-protocol:process-tx} for details of the specific operations).
 
-The number of constraints of \mimcMPPrime{} for several exponents $\exponent{}$ is given by the formula
-\[
-    \constraints = \rounds \cdot \mults + 1
-\]
-where $\rounds = \lceil \frac{\log_2 \rCURVE}{\log_2 \exponent{}} \rceil$, $\mults$ is the number of multiplications required for exponentiation and the additional constraint (corresponding to $+1$ in the above formula) is a result of the final message and key addition. Note that for $\exponent = 2^t - 1$ we have $\mults = 2 \cdot t - 2$, using the \emph{square-and-multiply} algorithm~\cite{menezes1996handbook}, and for $\exponent = 2^t + 1$ we have $mults = t + 1$.
-% TODO: Check if this final constraint can be removed.
+In the arithmetic circuit, an invocation of $\mimcMPPrime{}$ requires $\rounds \cdot \mults$ constraints, where $\mults$ is the number of multiplications required for exponentiation. For exponents of the form $\exponent = 2^t - 1$, we have $\mults = 2 \cdot t - 2$, (using the \emph{square-and-multiply} algorithm~\cite{menezes1996handbook}), and for $\exponent = 2^t + 1$ we have $mults = t + 1$. Thus we expect that exponents of the later form are more optimal.  In the contract, exponentiation is relatively much cheaper to execute than in an arithmetic circuit.  Therefore, given that the number of rounds decreases with higher $\exponent{}$, the gas cost of $\mimcMPPrime{}$ is also expected to decrease.
```

Comment:
> the gas cost of $\mimcMPPrime{}$ is also expected to decrease.

Let's try to be categorical when possible, instead of saying "it is expected to decrease". This can easily be verified with a small benchmark if necessary.

Reply:
I've never been sure about that sentence (I didn't write the original).

The idea of comparing costs between contract and circuit doesn't really make sense, (except maybe relative cost of multiplication vs some other op?) and idea that rounds decreasing increases efficiency applies similarly to all contexts.  So the intention was never very clear to me.  I've reworded it to just mention the circuit.

Reply:
> I've never been sure about that sentence (I didn't write the original).

Oh sorry, i thought you added it here!

> The idea of comparing costs between contract and circuit doesn't really make sense

I don't remember when that was written, but I don't think the goal is to compare circuit/contract complexity. The TL;DR is Zeth spans multiple execution environments:
1. Circuit computation
2. Contract (on-chain) computation
3. "normal" computation (on local computer)

Hence, to qualify a modification in the protocol as an "optimization" -> we need to take these 3 "worlds" into consideration. I.e. can we say that something that makes the proof generation faster is a "protocol optimization" if this makes the contract code prohibitively more expensive?
So the goal here AFAICT is to address this point precisely (not so much compare circuit and contract, but rather argue about the overall impact of an operation on the protocol perfs).

> and idea that rounds decreasing increases efficiency applies similarly to all contexts.

Reducing rounds numbers sounds like an optimization across the stack, but reducing the round number => increasing map degree which may impact perfs unheavenly in the different computational models. I think the question is: to which extend increased exponentiation cost outperforms the savings on the round reduction on the contract?

Reply:
> The TL;DR is Zeth spans multiple execution environmentsmodels where the costs aren't computed in the same way.
> ...

Hence I wanted to at least mention the contract cost in some way.  Even though the original points didn't seem quite right, I didn't want to remove it completely.

> Hence, to qualify a modification in the protocol as an "optimization" -> we need to take these 3 "worlds" into consideration.
> ...
> Hence, to qualify a modification in the protocol as an "optimization" -> we need to take these 3 "worlds" into consideration. I.e. can we say that something that makes the proof generation faster is a "protocol optimization" if this makes the contract code prohibitively more expensive?

We do have to consider those environments of course, but we are more finding e and r which are reasonably optimal over them, with various weightings.  Arguing about "optimizations" of the whole protocol seems somewhat out of scope.

> So the goal here AFAICT is to address this point precisely (not so much compare circuit and contract, but rather argue about the overall impact of an operation on the protocol perfs).
> Reducing rounds numbers sounds like an optimization across the stack, but reducing the round number => increasing map degree which may impact perfs unheavenly in the different computational models. I think the question is: to which extend increased exponentiation cost outperforms the savings on the round reduction on the contract?

Until now, the section talks about optimizing the number of constraints with only the original sentence to reference the contract.  To address the contract part more precisely might require working out the gas cost of the for-loop and some integer additions (i.e. the ops that the contract carries out which do not impact the number of constraints).  That would seem like overkill, and certainly out of scope for this change.  

For ref, the "original" I refer to is:

> Note that small exponents $\exponent{}$ result in fewer constraints in the arithmetic circuit (see~\cref{zeth-protocol:statement}), while larger exponents can reduce the cost of Merkle tree operations on the contract (see~\cref{zeth-protocol:process-tx}). This is due to two factors, namely that exponentiation is cheaper to execute on a contract than in an arithmetic circuit, and that the number of rounds decreases with higher $\exponent{}$.

It's probably the result of several iterations by various people, but the point is that it's not really valid/relevant now, and we could safely remove all mention of the contract cost without really impacating the section.

My intention was just to at least mention the contract cost, and to reason very simply: the cost on the contract will be dominated by the number of mulmods (exactly the number of circuit constraints), and so values of e and r for which this is low should also be (at least close to) optimal in the contract.  I've updated to try to make this clearer, but I may have misunderstood something.

Note, there may be some optimization we can do in the contract so that the performance characteristics change dramatically.  I did look for op codes when I was touching that code, but only found EXP (which may be useful if used with montgomery reduction? ... but that's probably for another change).  While the circuit and contract implementations work pretty much the same way, it seems sufficient to reason as above.

Let me know if that makes sense with the new version, or if I'm missing something here.

Reply:
>  Arguing about "optimizations" of the whole protocol seems somewhat out of scope.

Sure, I mean the same argument applies if we turn the sentence around and refer to an "adequate set of protocol parameters" instead of an "optimization", the point being, "to be an adequate set of protocol params, they need to yield efficient execution across the stack (i.e. on the contract, circuit and natively)".

>  To address the contract part more precisely might require working out the gas cost of the for-loop and some integer additions (i.e. the ops that the contract carries out which do not impact the number of constraints). That would seem like overkill, and certainly out of scope for this change.

Sure, the main point was: We need to be assertive in the specs, and avoid somehow speculative statements of the form "that should be fine" etc. So if we need extra data points to be more assertive (e.g. do benchmarks), we should definitely do it and add that as evidence in the appendix. If things are sufficiently clear without benchmark, absolutely no need to do some, that'd be overkill indeed.

> My intention was just to at least mention the contract cost, and to reason very simply: the cost on the contract will be dominated by the number of mulmods (exactly the number of circuit constraints), and so values of e and r for which this is low should also be (at least close to) optimal in the contract. I've updated to try to make this clearer, but I may have misunderstood something

Got it

> Let me know if that makes sense with the new version

LGTM, just approved the PR, thanks!